src/beamlime/constructors/providers.py

```diff
@@ -0,0 +1,155 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+from __future__ import annotations
+
+from functools import partial
+from types import FunctionType, LambdaType
+from typing import Callable, Dict, Literal, TypeVar, Union
+
+from .inspectors import (
+    DependencySpec,
+    ProductSpec,
+    ProductType,
+    collect_argument_dep_specs,
+)
+
+_Product = TypeVar("_Product")
+_lambda_name = (lambda: None).__qualname__
+
+
+def _validate_callable_as_provider(callable_obj: Constructor) -> Literal[True]:
+    if (
+        not isinstance(callable_obj, partial)
+        and hasattr(callable_obj, "__name__")
+        and not (callable_obj.__name__ == _lambda_name)
+        and hasattr(callable_obj, "__qualname__")
+        and callable_obj.__qualname__ != callable_obj.__name__
+    ):
+        raise NotImplementedError(
+            "A member method of a class can not be " "registered as a provider yet."
+        )
+    return True
+
+
+class Provider:
+    """
+    Function wrapper that provides certain type of product.
+
+    Nested ``Provider`` is avoided similar to ``partial``.
+
+    """
+
+    _constructor: Constructor
+    args: tuple
+    keywords: dict
+    product_spec: ProductSpec
+    arg_dep_specs: Dict[str, DependencySpec]
+
+    def __new__(
```

Comment:
Why did you need to use `__new__` here? This looks like it can be done in `__init__`.